lisp/ess-mode.el

```diff
@@ -989,6 +989,29 @@ generate the source buffer."
                   (down-list 1)
                 (error nil)))))))
 
+(defun ess-define-runner (ver dialect)
+  "Create a function VER.
+This function starts the inferior process with the specified
+version.  DIALECT can be \"R,\" \"S,\", \"SAS.\""
+  (let ((name (intern ver))
+        (body (cond ((string= dialect "R")
+                     `(let ((inferior-R-version ,ver)
+                            (inferior-ess-r-program-name (if ess-microsoft-p
+                                                             "Rterm" "R")))
+                        (R start-args)))
+                    ((string= dialect "S")
+                     `(let ((inferior-S+-program-name ,ver))
+                        (S+)))
+                    ((string= dialect "SAS")
+                     `(let ((inferior-SAS-program-name ,ver))
+                        (SAS start-args))))))
+    (eval
```

Comment:
I think this would be cleaner with `fset` and a function that closes over `name` and `ver` (unless I'm missing something).

Reply:
In general with a function it's more natural to wrap `fset` and with a macro it's more natural to wrap `defun`.

Reply:
Do you mean something like this (untested)?

```emacs-lisp

(defun ess-define-runner (ver dialect)
  "Create a function VER.
This function starts the inferior process with the specified
version.  DIALECT can be \"R,\" \"S,\", \"SAS.\""
  (fset ,(intern ver)
        (lambda (&optional start-args)
          "Start this process version in an inferior ESS buffer.
Function defined using `ess-define-runner'."
          (cond ((string= dialect "R")
                 `(let ((inferior-R-version ,ver)
                        (inferior-ess-r-program-name (if ess-microsoft-p
                                                         "Rterm" "R")))
                    (R start-args)))
                ((string= dialect "S")
                 `(let ((inferior-S+-program-name ,ver))
                    (S+)))
                ((string= dialect "SAS")
                 `(let ((inferior-SAS-program-name ,ver))
                    (SAS start-args)))))))
```

Reply:
Could this work, given that `ver` and `dialect` are not known at compile-time? I was thinking about using `(lexical-let ((ver ver) (dialect dialect)) (lambda ...))` and then you don't need any expansion / unquoting.